src/Microsoft.OData.Edm/Csdl/Semantics/CsdlSemanticsOperation.cs

```diff
@@ -132,7 +134,114 @@ private IEnumerable<IEdmOperationParameter> ComputeParameters()
+            // If we find at least one of out-of-line optional parameter annotation, we replace it as an optional parameter.
+            // Otherwise, re-use the built parameter.
+            // Be noted: if a parameter has inline and out-of-line optional parameter, the out-of-line will win.
+            string fullName = ParameterizedTargetName(Namespace + "." + Name, parameters);
+
+            List<IEdmOperationParameter> newParameters = new List<IEdmOperationParameter>();
+            foreach (var parameter in parameters)
+            {
+                // for example: NS.TestFunction(Edm.String, Edm.String, Edm.String)/OptionalParameter
+                string fullTargetName = fullName + "/" + parameter.Name;
+                string defaultValue;
+
+                if (TryGetOptionalParameterAnnotation(fullTargetName, out defaultValue))
+                {
+                    CsdlSemanticsOperationParameter csdlSemanticsParameter = (CsdlSemanticsOperationParameter)parameter;
+                    newParameters.Add(new CsdlSemanticsOptionalParameter(this, (CsdlOperationParameter)csdlSemanticsParameter.Element, defaultValue));
+                }
+                else
+                {
+                    newParameters.Add(parameter);
+                }
+            }
+
+            return newParameters;
+        }
+
+        private bool TryGetOptionalParameterAnnotation(string fullTargetName, out string defaultValue)
+        {
+            defaultValue = null;
+            bool isOptional = false;
+
+            List<CsdlSemanticsAnnotations> annotations;
+            if (Model.OutOfLineAnnotations.TryGetValue(fullTargetName, out annotations))
+            {
+                foreach (var annotation in annotations)
+                {
+                    var optionalParameterAnnotation = annotation.Annotations.Annotations.FirstOrDefault(a =>
+                        a.Term == CoreVocabularyModel.OptionalParameterTerm.ShortQualifiedName() ||
+                        a.Term == CoreVocabularyModel.OptionalParameterTerm.FullName());
+
+                    if (optionalParameterAnnotation != null)
+                    {
+                        isOptional = true;
+
+                        CsdlRecordExpression optionalValueExpression = optionalParameterAnnotation.Expression as CsdlRecordExpression;
+                        if (optionalValueExpression != null)
+                        {
+                            foreach (CsdlPropertyValue property in optionalValueExpression.PropertyValues)
+                            {
+                                CsdlConstantExpression propertyValue = property.Expression as CsdlConstantExpression;
+                                if (propertyValue != null)
+                                {
+                                    if (property.Property == CsdlConstants.Attribute_DefaultValue)
+                                    {

```

Comment:
should probably first check to see if property.Property == CsdlConstants.Attribute_DefaultValue, and then try to cast property.Expression to CsdlConstantExpression (since the comparison is cheaper than the cast).  i.e., move 192-193 before 189.

Reply:
Resolved.